presto-hive/src/main/java/io/prestosql/plugin/hive/security/SqlStandardAccessControlMetadata.java

```diff
@@ -84,6 +88,69 @@ private static void checkRoleIsNotReserved(String role)
         return ImmutableSet.copyOf(metastore.listRoles());
     }
 
+    @Override
+    public Set<RoleGrant> listAllRoleGrants(ConnectorSession session, Optional<Set<String>> roles, Optional<Set<String>> grantees, OptionalLong limit)
+    {
+        /*
+         * The Hive Metadata interface provides us with two ways to retrieve role grants:
+         * 1. principals by single role
+         * 2. roles by single principal
+         *
+         * We check the passed predicates to decide which path is most optimal.
+         */
+        if (grantees.isPresent()) {
+            if (roles.isPresent() && roles.get().size() <= grantees.get().size() * 2) {
+                // 2x because for the grantees we do not know the grantee type (ROLE or USER)
+                return getRoleGrantsByRoles(roles.get(), limit);
+            }
+            else {
+                return getRoleGrantsByGrantees(grantees.get(), limit);
+            }
+        }
+        else {
+            // if roles are present use those, otherwise get a list of roles from the metastore
+            return getRoleGrantsByRoles(roles.orElseGet(() -> metastore.listRoles()), limit);
+        }
```

Comment:
```suggestion
        Set<String> actualRoles = roles.orElseGet(() -> metastore.listRoles());
      
        // choose more efficient path
        if (grantees.isPresent() && actualRoles.size() > grantees.get().size() * 2) {
                // 2x because we check two grantee types (ROLE or USER)
                return getRoleGrantsByGrantees(grantees.get(), limit);
        }
        return getRoleGrantsByRoles(actualRoles, limit);
        
```

Reply:
That leads to a possibly unnecessary call to the HMS, though.

Reply:
On the other hand it's a nice safety check to upper bind the number to RPC to no more that Roles+1. So I'll do that indeed.